DDF-1555: Implement solution for cardinality issues that cause unpredictable configuration

lessarderic opened the pull request:
Design overview: https://codice.atlassian.net/secure/attachment/12101/Configuration%20Design.JPG

Reviewers:
@oconnormi 
@garrettfreibott 
@figliold 
@brendan-hofmann
@pklinef 

<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/codice/ddf/305)
<!-- Reviewable:end -->

brendan-hofmann commented:
Reviewed 8 of 32 files at r1, 6 of 20 files at r2, 11 of 13 files at r3, 6 of 9 files at r4, 5 of 5 files at r5.
Review status: all files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java, line 14 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28J5fASqpkaoJCJV5Z)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java#L14)):</sup>
I think it's about time to rename this test package, we have a lot more than catalog testing in here now. ;)
<!-- Reviewable comment -K28J5fASqpkaoJCJV5_:971452850 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java, line 55 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28JjPqjIVpIft4T0-N)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java#L55)):</sup>
Won't we want some of these utility classes extracted eventually so other itests can leverage them? At least the configuration editing in AbstractIntegrationTest? We might also consider pre-configuring test loopback sources and the like in this manner.
<!-- Reviewable comment -K28JQ61vBV_zQDIvDMQ:-1669991356 -->

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 532 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28LILKjutaq0VOmil7)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-app/src/main/resources/features.xml#L532)):</sup>
We don't use `start-level`very much, is it significant here? If so, a comment to that effect might be a good idea.
<!-- Reviewable comment -K28LILKjutaq0VOmil8:-244035320 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 56 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28M01DYU1ASE494-Qj)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/pom.xml#L56)):</sup>
I think it's about time we homogenized our versions for guava, junit, and hamcrest.
<!-- Reviewable comment -K28M01ECa9ssJmK4iep:-455855452 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 36 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28NwmgSbn_aNJ26Vcg)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L36)):</sup>
I can't find documentation one way or another, but I recall that at least in Java 7, filesystem events were somewhat inconsistent in their behavior between different filesystems and OSes. How many OSes/filesystems has this been tested on, or does Java 8 offer stronger guarantees of consistency?
<!-- Reviewable comment -K28NwmhiChMwwKGJ18-:242989781 -->

---

<sup>**[platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 47 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28PU_TCZELuqjQrkwP)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L47)):</sup>
It looks like there are some potential gotchas wrt escape characters and path separators on different OSes/filesystems here. The system property approach is fine, but we'll want to document clearly what those implications are. (thinking particularly of windows here...)
<!-- Reviewable comment -K28PU_Ubhw1MH_lfBoL:-1029099952 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java, line 14 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28J5fASqpkaoJCJV5Z)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java#L14)):</sup>
Agreed. Created DDF-1621.
<!-- Reviewable comment -K2D97S-UpreVCPQweJ3:280780348 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: 34 of 35 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java, line 55 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28JjPqjIVpIft4T0-N)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java#L55)):</sup>
That's a good point. Some of it will already be available to everyone, e.g., `ConfigurationPropertiesEqualTo` hamcrest `Matcher` and `AbstractIntegrationTest.installStartupFile()`. The rest of that class is pretty much specific to these tests and I'm not sure how much of it would be useful outside. Especially since our goal is to replace all the manual configurations we have today in our itests and use `.config` files instead to pre-configure the tests. But if you feel there's something in there that might be useful, I'm always open to extracting common code :-)
<!-- Reviewable comment -K2D9_CWAiMltCrGsAm0:698723059 -->

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 532 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28LILKjutaq0VOmil7)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-app/src/main/resources/features.xml#L532)):</sup>
Good question. @figliold, since the feature is first in `platform-app`, do we even need to give it a `start-level`? If so, I would actually have expected it to be lower than 55.
<!-- Reviewable comment -K2DD5sWR4A3Tc2nfyPt:-1052762537 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 56 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28M01DYU1ASE494-Qj)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/pom.xml#L56)):</sup>
We already have a ticket for this.
<!-- Reviewable comment -K2DDT-DrNNBeaW8dKq1:-423780985 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 36 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28NwmgSbn_aNJ26Vcg)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L36)):</sup>
We did notice fairly long delays with this in our first implementation, when we cared about updates and deletes, which made our itests somewhat inconsistent. But since we're only watching for file creation, we felt it didn't matter as much and decided to keep using this. We can always switch to something better in the future if we ever need to watch for updates and deletes again. Thoughts?
<!-- Reviewable comment -K2DDbCsBhKWKmb8NvgE:1042017066 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1179 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DInZAdym52Gq6gSHp)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1179)):</sup>
I'm fine with phrasing the documentation this way, but technically the 'undefined' behavior is just a race condition, right?
<!-- Reviewable comment -K2DInZAdym52Gq6gSHq:-588668494 -->

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1195 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DJ1ED48IpjzdJXpWD)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1195)):</sup>
typo: Manage should be Managed?
<!-- Reviewable comment -K2DJ1ED48IpjzdJXpWE:-1300483172 -->

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1199 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DJB3-rJum59I_102Y)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1199)):</sup>
Is this sequential number just for disambiguation on the filesystem, or does it have behavioral implications as well? Does it affect what order the instances are created in? Do you need to know how  many instances already exist to know what number to use?
<!-- Reviewable comment -K2DJB300eBQrD3qabdX:1876955468 -->

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1211 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DJVBOgbEL9W8rzoeh)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1211)):</sup>
Perhaps mention that this is also defined in blueprint/spring, at least for our modules? Or are we sure all our services have metatypes?
<!-- Reviewable comment -K2DJVBOgbEL9W8rzoei:-559674998 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java, line 55 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28JjPqjIVpIft4T0-N)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java#L55)):</sup>
I know there are itests that programmatically modify configurations right now, I guess I was just wondering if this would be an easier way to do that. It is a bit weird to go through the filesystem for that, but I was thinking it might be useful for eventually doing real federated itests. I guess we can cross that bridge when we come to it, I don't know if we have enough specifics to address it yet. ;)
<!-- Reviewable comment -K2DKchse7VL4IU_y5Bt:1303739124 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 56 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28M01DYU1ASE494-Qj)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/pom.xml#L56)):</sup>
Yeah, that occurred to me later, my bad.
<!-- Reviewable comment -K2DL0JUmexfR-GAPr_a:1900732814 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 36 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28NwmgSbn_aNJ26Vcg)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L36)):</sup>
I did some more googling and think I found some of the limitations I was thinking of. They mostly involved being stuck with polling solutions and only being able to get directory (not file) events of certain types on certain oses/filesystems. It looks to me like our usage should be unaffected though?
<!-- Reviewable comment -K2DLa0iYMR_LtCGfVhw:1490372569 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

garrettfreibott commented on platform/platform-configuration-listener/pom.xml:
```diff
@@ -0,0 +1,144 @@
+                            <rules>
+                                <rule>
+                                    <element>BUNDLE</element>
+                                    <!-- Because of some JaCoCo limitations related to
```
Is this still an issue? The limits seem to be set higher than 75% here.

lessarderic commented:
Review status: 33 of 36 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1179 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DInZAdym52Gq6gSHp)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1179)):</sup>
Yes. We just don't want them to have both :-)
<!-- Reviewable comment -K2J18Ob-UEynDJ_4plY:-1128911164 -->

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1195 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DJ1ED48IpjzdJXpWD)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1195)):</sup>
Good catch! Fixed.
<!-- Reviewable comment -K2J2T4HNnyY7xKe47i1:1899284406 -->

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1211 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DJVBOgbEL9W8rzoeh)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1211)):</sup>
I guess my worry here is that our blueprint files are more internal than metatype files. If a configuration files doesn't have a corresponding metatype file, it usually is not really meant to be configured externally using a UI. We would also have to explain how to create a valid config file from a blueprint config and potentially find a way to generate template files (which is a follow on ticket) from those blueprint files, which is much more difficult than generating them from metatype files.
<!-- Reviewable comment -K2J8LORYVO4eLBYgz1w:-385595440 -->

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java, line 55 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28JjPqjIVpIft4T0-N)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java#L55)):</sup>
Sounds good to me.
<!-- Reviewable comment -K2J9V41G2aZhDEhqUzc:-682275779 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 56 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28M01DYU1ASE494-Qj)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/pom.xml#L56)):</sup>
No worries :-)
<!-- Reviewable comment -K2J9Xvc824VVVh2fKIj:-1012227845 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 107 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/305#-K1vC7EDRn8YA9d26Ll7-r8-107)** ([raw file](https://github.com/codice/ddf/blob/c589e74d2bd3a3ba80afa861fe5e71fc7668094d/platform/platform-configuration-listener/pom.xml#L107)):</sup>
Removed.
<!-- Reviewable comment -K2J9n0xx4bWYP0BZAMx:1740281843 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 36 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28NwmgSbn_aNJ26Vcg)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L36)):</sup>
Did you keep those links? We do care about file creation events (not directories) so I want to make sure it doesn't affect us.
<!-- Reviewable comment -K2J9ueqShhfF8dCWmdU:-795929254 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Reviewed 3 of 5 files at r8.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1211 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DJVBOgbEL9W8rzoeh)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1211)):</sup>
Although I personally insist on being able to shoot myself in the foot if I want, I can understand not mentioning it in our documentation. sgtm
<!-- Reviewable comment -K2JLOJCe_VNHmqYzuZj:520466348 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 107 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/305#-K1vC7EDRn8YA9d26Ll7-r8-107)** ([raw file](https://github.com/codice/ddf/blob/c589e74d2bd3a3ba80afa861fe5e71fc7668094d/platform/platform-configuration-listener/pom.xml#L107)):</sup>
The argument for specifying limits higher than 75% is so that we don't regress unknowingly. I personally prefer to have the higher limits there, since one long-term goal is to automatically keep those numbers up-to-date so we always push higher. ;)
<!-- Reviewable comment -K2JJj_Q4DnEJtNFbEwH:-1110906174 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 36 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28NwmgSbn_aNJ26Vcg)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L36)):</sup>
It wasn't anything very authoritative, I'm afraid. The official docs only have the following to say:

*Platform Dependencies*

The implementation that observes events from the file system is intended to map directly on to the native file event notification facility where available, or to use a primitive mechanism, such as polling, when a native facility is not available. Consequently, many of the details on how events are detected, their timeliness, and whether their ordering is preserved are highly implementation specific. For example, when a file in a watched directory is modified then it may result in a single ENTRY_MODIFY event in some implementations but several events in other implementations. Short-lived files (meaning files that are deleted very quickly after they are created) may not be detected by primitive implementations that periodically poll the file system to detect changes.

If a watched file is not located on a local storage device then it is implementation specific if changes to the file can be detected. In particular, it is not required that changes to files carried out on remote systems be detected.
<!-- Reviewable comment -K2JK0Y4weDEGmAEBQgx:-1619584867 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

<sup>**[distribution/docs/src/main/resources/ManContents.adoc, line 1199 \[r7\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2DJB3-rJum59I_102Y)** ([raw file](https://github.com/codice/ddf/blob/8ea26a2cc345184c737c9a2ed2010a0095f43954/distribution/docs/src/main/resources/ManContents.adoc#L1199)):</sup>
Clarified how the number is being used.
<!-- Reviewable comment -K2JURvbChTEv16AqTmz:1296477374 -->

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 107 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/305#-K1vC7EDRn8YA9d26Ll7-r8-107)** ([raw file](https://github.com/codice/ddf/blob/c589e74d2bd3a3ba80afa861fe5e71fc7668094d/platform/platform-configuration-listener/pom.xml#L107)):</sup>
I didn't remove the whole thing. Just the comment. So we _do_  have thresholds higher than 75% now :-)
<!-- Reviewable comment -K2JRkFOBSCe6e-kFiBv:-392961728 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 36 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28NwmgSbn_aNJ26Vcg)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L36)):</sup>
Looks like we should be good for now. But we'll keep that in mind if we ever do more from a file watch perspective. Thanks!
<!-- Reviewable comment -K2JS3QETwjUJEQ0Q8YD:-1560446012 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

figliold commented:
:thumbsup: 
<!-- Reviewable comment -K2JWs-zbzXqAthAWX-n:-453412105 -->

---

Review status: 34 of 36 files reviewed at latest revision, 3 unresolved discussions.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Reviewed 2 of 2 files at r9.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 107 \[r8\]](https://reviewable.io:443/reviews/codice/ddf/305#-K1vC7EDRn8YA9d26Ll7-r8-107)** ([raw file](https://github.com/codice/ddf/blob/c589e74d2bd3a3ba80afa861fe5e71fc7668094d/platform/platform-configuration-listener/pom.xml#L107)):</sup>
:+1: 
<!-- Reviewable comment -K2JVqCCTjr3WoIJH1gY:-510466975 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java, line 36 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28NwmgSbn_aNJ26Vcg)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFilesPoller.java#L36)):</sup>
It would be worth actually testing on different OSes/filesystems someday, just in case for things like this. ;)
<!-- Reviewable comment -K2JWwD-qJo7MY4EWXQd:-1693155298 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Reviewed 1 of 32 files at r1, 1 of 20 files at r2, 10 of 13 files at r3, 1 of 5 files at r5, 3 of 5 files at r8, 1 of 2 files at r9.
Review status: all files reviewed at latest revision, 5 unresolved discussions.

---

<sup>**[distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java, line 268 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2Ja3GD4VB548Lj6Mac)** ([raw file](https://github.com/codice/ddf/blob/9bd2967fc121bb8572772977c4fba8bd2946b8a2/distribution/test/itests/test-itests-catalog/src/test/java/ddf/catalog/test/TestPlatform.java#L268)):</sup>
Need to add this.
<!-- Reviewable comment -K2Ja3GD4VB548Lj6Mad:488671857 -->

---

<sup>**[platform/platform-app/src/main/resources/features.xml, line 532 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28LILKjutaq0VOmil7)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-app/src/main/resources/features.xml#L532)):</sup>
Based on the Karaf documentation and the feature schema, `start-level` applies only to `<bundle> elements, not `<feature>` elements. I'll remove it.
<!-- Reviewable comment -K2Jd1rbvSx2ApXDz3yI:-2039543540 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Reviewed 3 of 20 files at r2, 1 of 5 files at r5, 1 of 2 files at r9.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 27 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2JduuYNegLjfztM3ov)** ([raw file](https://github.com/codice/ddf/blob/9bd2967fc121bb8572772977c4fba8bd2946b8a2/platform/platform-configuration-listener/pom.xml#L27)):</sup>
Update description.
<!-- Reviewable comment -K2JduuZfT87l-p7DlgP:1116513380 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ChangeListener.java, line 29 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2JeUKhc0VIMHBnmdzV)** ([raw file](https://github.com/codice/ddf/blob/9bd2967fc121bb8572772977c4fba8bd2946b8a2/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ChangeListener.java#L29)):</sup>
Update JavaDoc.
<!-- Reviewable comment -K2JeUKhc0VIMHBnmdzW:1572640010 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFile.java, line 26 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2JfD1-IGCmvgjzFFQi)** ([raw file](https://github.com/codice/ddf/blob/9bd2967fc121bb8572772977c4fba8bd2946b8a2/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFile.java#L26)):</sup>
Unused. Remove.
<!-- Reviewable comment -K2JfD1-IGCmvgjzFFQj:680247613 -->

---

<sup>**[platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFileFactory.java, line 34 \[r9\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2JgykrSzrNHls82tXl)** ([raw file](https://github.com/codice/ddf/blob/9bd2967fc121bb8572772977c4fba8bd2946b8a2/platform/platform-configuration-listener/src/main/java/org/codice/ddf/configuration/store/ConfigurationFileFactory.java#L34)):</sup>
Need JavaDoc
<!-- Reviewable comment -K2JgykrSzrNHls82tXm:-1123260495 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Reviewed 5 of 32 files at r1, 2 of 20 files at r2, 1 of 13 files at r3, 6 of 9 files at r4, 2 of 5 files at r5.
Review status: all files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Reviewed 13 of 14 files at r10, 1 of 1 files at r11.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 27 \[r11\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2NCKRTR0TCQjWQ5y0k)** ([raw file](https://github.com/codice/ddf/blob/04ee440abd84223ea528d59804fdbc1023914b5e/platform/platform-configuration-listener/pom.xml#L27)):</sup>
This description implies that we don't listen for config files during runtime as well, which the documentation still claims we do. Can we align the description and the documentation better?
<!-- Reviewable comment -K2NCKRTR0TCQjWQ5y0l:-442148470 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Reviewed 1 of 1 files at r11.
Review status: 36 of 37 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-configuration-listener/pom.xml, line 27 \[r11\]](https://reviewable.io:443/reviews/codice/ddf/305#-K2NCKRTR0TCQjWQ5y0k)** ([raw file](https://github.com/codice/ddf/blob/04ee440abd84223ea528d59804fdbc1023914b5e/platform/platform-configuration-listener/pom.xml#L27)):</sup>
Updated.
<!-- Reviewable comment -K2OBM8pXk-62MqdTVg3:-719267080 -->

---

<sup>**[platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 47 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28PU_TCZELuqjQrkwP)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L47)):</sup>
Trying to build and run the itests on Windows to make sure there are no OS specific issues with our code. Will update PR if needed.

I'm also working on a Bamboo job to run the itests on Windows. It will only build `master` but we will at least have one easy to run the itests if needed and we can always make that part of our nightly builds eventually.
<!-- Reviewable comment -K2DGZ3oJ4sIla8Mb2h3:-1681157525 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

lessarderic commented:
Reviewed 13 of 14 files at r10.
Review status: 36 of 37 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
Reviewed 1 of 1 files at r12.
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

<sup>**[platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml, line 47 \[r5\]](https://reviewable.io:443/reviews/codice/ddf/305#-K28PU_TCZELuqjQrkwP)** ([raw file](https://github.com/codice/ddf/blob/ef2f7a2b3f9bf1d4c5b848bc48e1e5d519641e2c/platform/platform-configuration-listener/src/main/resources/OSGI-INF/blueprint/blueprint.xml#L47)):</sup>
I dream of it being incorporated into the PR build as well, but having it in our nightlies is a very good start. :thumbsup: 
<!-- Reviewable comment -K2O_FiMVgERHA-NnJuP:-1661091656 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/codice/ddf/305)
<!-- Sent from Reviewable.io -->

brendan-hofmann commented:
![selfie-0](http://i.imgur.com/l2j4yzT.png)
